sdks/python/apache_beam/internal/pickler.py

```diff
@@ -40,9 +40,22 @@
 
 import dill
 
-# Pickling, especially unpickling, can cause broken module imports
-# if executed concurrently, see: BEAM-8651.
-pickle_lock = threading.Lock()
+
+class _NoOpContextManager(object):
+  def __enter__(self):
+    pass
+
+  def __exit__(self, *unused_exc_info):
+    pass
+
+
+if sys.version_info[0] > 2:
+  # Pickling, especially unpickling, causes broken module imports on Python 3
+  # if executed concurrently, see: BEAM-8651, http://bugs.python.org/issue38884.
+  pickle_lock_unless_py2 = threading.RLock()
```

Comment:
Have you mentioned perf issues related to RLock? If yes, should we add a link to that?

Reply:
It's supposedly only below Python 3.3: https://bugs.python.org/issue3001 (currently the link is in PR).